### ISSUE TITLE
Inference float16 support

### ISSUE THREAD

**Issue by kexinzhao** [PaddlePaddle/Paddle#8693]
Below are the list of to-do items for supporting float16 inference.

- [x] make float16 a POD type (#8456)
- [x] data_type_transform for fp16 (#8619 #8850)
- [x] add fp16 gemm function into math_function (#8695)
- [x] add fp16 support for mul_op (#8817)
- [x] bind float16 to numpy.float16
- [x] add fp16 for cudnn conv2d_op 
- [x] add fp16 for cudnn pool2d_op
- [x] cast op
- [x] batch norm op
- [x] dropout
- [x] use tensor core for cublas
- [x] use tensor core for cudnn
- [x] have a vgg16 cifar10 inference python example that runs on float32 mode
- [x] have a vgg16 cifar10 inference python example that runs on float16 mode
- [x] get some initial result (verify tensor cores works, verify fp16 is faster, verify accuracy is comparable
- [x] relu
- [x] element-add op
- [x] softmax
- [x] tanh and other activations
- [x] float16 example python
- [x] float16 example c++ inference
- [x] float16 example correctness
- [x] float16 transpiler and example
- [x] update float16 design doc
- [x] save
- [x] load
- [x] float16 correctness code
- [x] float16 benchmark code
- [x] float16 report in English
- [x] conv_3d
- [x] pool_3d

Future:
- [ ] sum op
- [ ] other element-wise operator (sub divide mul min max pow)
- [ ] cudnn relu
